ls1intum/Artemis
# `Quiz`: Implement quiz participation

## Summary
Adds the initial student-facing quiz participation workflow. Quiz exercises in the course exercise list now open a quiz participation route where students can answer multiple-choice questions, see score, remaining time, and last-submission information, and submit answers while the quiz is still active.

## Motivation and Context
This enables students to participate in quiz exercises from the course page instead of only creating or managing quiz content. No issue is linked to this PR.

Original note from the PR: The Submit button behavior differs from what was discussed in the team call, and the socket connection is still missing.

## Description
- Adds the AngularJS quiz participation route, controller, and template under `src/main/webapp/app/quiz/participate/`, wires them in `index.html`, and updates quiz styles, labels, and English/German translations.
- Updates the course exercise list so quiz exercises navigate to `/quiz/{id}`, use quiz-specific start/resume text and status states, and rely on the backend `isVisibleToStudents` flag for release visibility.
- Extends the multiple-choice participation component so selected answer options are managed by the parent quiz controller, compared by answer option id, and displayed together with the question score.
- Adds student-safe quiz exercise retrieval via `/api/quiz-exercises/{id}/for-student`, filtering explanations and correct-answer markers from the response.
- Changes course exercise retrieval to filter by course permissions and student visibility on the backend, and removes quiz questions from exercise-list payloads.
- Adds quiz submission support with `/courses/{courseId}/exercises/{exerciseId}/submissions/my-latest` to initialize or retrieve the current participation/result/submission and `PUT /api/quiz-submissions` to save submitted answers and update the result completion date before the quiz ends.
- Extends quiz submission and submitted-answer persistence with a submitted-answer-to-question relation, JSON subtype metadata for submitted answer types, cascading eager quiz submitted answers, a transient submission date, repository helpers, Liquibase migration, and updated tests for the changed resource constructor.

## Steps for Testing
Prerequisites:
- A course with a planned quiz exercise containing multiple-choice questions.
- One instructor or teaching assistant and one student enrolled in the course.

1. Log in as the student and open the course exercise list after the quiz is visible to students.
2. Verify that the quiz row shows quiz-specific status text and Start Quiz or Resume Quiz, then start the quiz.
3. Verify that `/quiz/{exerciseId}` opens and shows the course title, quiz title, questions, question scores, total score, remaining time, and last-submission information.
4. Select and deselect multiple-choice answer options, submit the quiz, and verify that the last-submission display updates.
5. Reload or resume the quiz and verify that the previous selected answer options are loaded from the latest submission.
6. Verify that student-facing quiz responses do not expose correct-answer markers or explanations in the network payload.
7. Let the quiz time expire or use a quiz whose due date is in the past, then verify that submitting is disabled or rejected.
8. As instructor or teaching assistant, verify that the exercise list still shows unreleased exercises as not released, while students only receive exercises visible to them.

## Checklist
- [x] I have read and followed the [coding guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/)
- [x] PR title follows the Artemis naming convention.
- [x] PR description matches the actual diff.
- [ ] Manual quiz participation flow has been tested on a suitable test setup.
- [ ] Relevant automated tests have been run.
